File: gosource/signal/fabo/fabostrat.go
package fabo

import (
	"global"
	"math"
)

type FaboStrat struct {
	tp          float64
	sl          float64
	tole        float64
	Signal_list []global.Signal_obj
	Profit_dict map[string]Profit_obj
}

type Profit_obj struct {
	Open_price  float64
	Close_time  string
	Close_price float64
	Profit      float64
	Profit_rate float64
	Tp_rate     float64
	Sl_rate     float64
	Type        int
}

func (F *FaboStrat) Init() {
	F.tp = 0.382
	F.sl = 0.618
	F.tole = 0.001
	F.Signal_list = make([]global.Signal_obj, 0)
	F.Profit_dict = make(map[string]Profit_obj, 0)
}

func (F *FaboStrat) Profit_tick(signal_dict global.Signal_obj, price float64, datetime string) Profit_obj {
	// fmt.Println(signal_dict)
	cur_signal := signal_dict
	hl0 := cur_signal.P1_price
	hl1 := cur_signal.P2_price
	hl2 := cur_signal.P3_price
	crtype := cur_signal.Type
	delta0 := math.Abs(hl0 - hl1)
	delta1 := math.Abs(hl1 - hl2)
	tp_rate, sl_rate := 0., 0.
	delta2 := 0.
	if crtype == 0 {
		delta2 = hl2 - price
		if delta2 < 0 {
			delta3 := hl1 - price
			sl_rate = delta3 / delta0
		} else {
			tp_rate = delta2 / delta1
		}
	} else if crtype == 1 {
		delta2 = price - hl2
		if delta2 < 0 {
			delta3 := price - hl1
			sl_rate = delta3 / delta0
		} else {
			tp_rate = delta2 / delta1
		}
	}
	if ((tp_rate > F.tp) || (-sl_rate > F.sl)) && math.Abs(hl0-hl1)/hl0 > 0.001 {
		temp_profit_obj := Profit_obj{Open_price: hl2, Close_time: datetime, Close_price: price, Profit: delta2, Profit_rate: delta2 / hl2, Tp_rate: tp_rate, Sl_rate: sl_rate, Type: crtype}

		return temp_profit_obj
	} else {
		temp_profit_obj := Profit_obj{Close_time: "-1"}
		return temp_profit_obj
	}

}

func (F *FaboStrat) Profit(signal_dict global.Signal_obj, price_dict global.Price_dic) {
	if len(F.Signal_list) != 0 {
		if F.Signal_list[len(F.Signal_list)-1].P3_time != signal_dict.P3_time {
			temp_dic := signal_dict
			if temp_dic.P3_time != "" {
				F.Signal_list = append(F.Signal_list, temp_dic)
			}
		}
		for i := 0; i < len(F.Signal_list); i++ {
			cur_signal := F.Signal_list[i]
			crtime := cur_signal.P3_time
			_, ok := F.Profit_dict[crtime]
			if ok {
				continue
			}
			hl0 := cur_signal.P1_price
			hl1 := cur_signal.P2_price
			hl2 := cur_signal.P3_price
			crtype := cur_signal.Type
			delta0 := math.Abs(hl0 - hl1)
			delta1 := math.Abs(hl1 - hl2)
			tp_rate, sl_rate := 0., 0.
			delta2 := 0.
			if crtype == 0 {
				delta2 = hl2 - price_dict.Close
				if delta2 < 0 {
					delta3 := hl1 - price_dict.Close
					sl_rate = delta3 / delta0
				} else {
					tp_rate = delta2 / delta1
				}
			} else if crtype == 1 {
				delta2 = price_dict.Close - hl2
				if delta2 < 0 {
					delta3 := price_dict.Close - hl1
					sl_rate = delta3 / delta0
				} else {
					tp_rate = delta2 / delta1
				}
			}
			if (tp_rate > F.tp) || (-sl_rate > F.sl) {
				temp_profit_obj := Profit_obj{Open_price: hl2, Close_time: price_dict.Date_time, Close_price: price_dict.Close, Profit: delta2, Profit_rate: delta2 / hl2, Tp_rate: tp_rate, Sl_rate: sl_rate, Type: crtype}
				F.Profit_dict[crtime] = temp_profit_obj
				// fmt.Println(temp_profit_obj.Profit)
			}
		}
	} else {
		temp_dic := signal_dict
		if temp_dic.P3_time != "" {
			F.Signal_list = append(F.Signal_list, temp_dic)
		}
	}
}

File: pysource/Strategies/fabostrat.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
# LUX et VERITAS
# Create On: 2022/12/05 13:26:14



class FaboStrat(object):
    
    def __init__(self, tp=0.382, sl=0.618, tole=0.001):
        """
        Parameters
        ------------
            tp : float
                take profit
            sl : float
                stop-losses
        """
        
        self._tp = tp
        self._sl = sl
        self._profit_dict = dict()
        self._signal_list = list()
        
    
    def profit(self, signal_dict, price_dict: dict=dict()):
        """
        
        """
        
        if len(self._signal_list):
            if self._signal_list[-1] != signal_dict:
                self._signal_list.append(signal_dict)
            del_list = list()
            for cur_idx, cur_signal in enumerate(self._signal_list):
                crtime = cur_signal["p3_time"]
                hl0 = cur_signal["p1_price"]
                hl1 = cur_signal["p2_price"]
                hl2 = cur_signal["p3_price"]
                crtype = cur_signal["type"]
                delta0 = abs(hl0 - hl1)
                delta1 = abs(hl1 - hl2)
                tp_rate, sl_rate = 0, 0
                
                if crtype == 0:
                    delta2 = hl2 - price_dict["close"]
                    if delta2 < 0:
                        delta3 = hl1 - price_dict["close"]
                        sl_rate = delta3/delta0
                    else:
                        tp_rate = delta2/delta1
                elif crtype == 1:
                    delta2 = price_dict["close"] - hl2
                    if delta2 < 0:
                        delta3 = price_dict["close"] - hl1
                        sl_rate = delta3/delta0
                    else:
                        tp_rate = delta2/delta1
                
                if ((tp_rate > self._tp) or (-sl_rate > self._sl)) and (not crtime in self._profit_dict):
                    self._profit_dict[crtime] = dict()
                    # 开仓价格
                    self._profit_dict[crtime]["open_price"] = hl2
                    # 平仓时间
                    self._profit_dict[crtime]["close_time"] = price_dict["date_time"]
                    # 平仓价格
                    self._profit_dict[crtime]["close_price"] = price_dict["close"]
                    # 收益金额
                    self._profit_dict[crtime]["profit"] = delta2
                    # 收益比例（不带杠杆）
                    self._profit_dict[crtime]["profit_rate"] = delta2/hl2
                    # 止盈点比例（FABO）
                    self._profit_dict[crtime]["tp_rate"] = tp_rate
                    # 止损点比例（FABO）
                    self._profit_dict[crtime]["sl_rate"] = sl_rate
                    # 交易类型（做多/做空）
                    self._profit_dict[crtime]["type"] = crtype
                    
                    if cur_idx < len(self._signal_list)-2:
                        del_list.append(cur_idx)
                if crtime in self._profit_dict:
                    if cur_idx < len(self._signal_list)-2:
                        del_list.append(cur_idx)

            for i in del_list:
                self._signal_list.pop(i)
            
        else:
            self._signal_list.append(signal_dict.copy())